# Tarea 1 Bases de Datos 2023-1 Grupo XX

_Recuerde modificar esta celda y la siguiente! (Borre esta línea también)._

## Integrantes:
* Nombre Integrante 1 - Rol-1
* Nombre Integrante 2 - Rol-2

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [102]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names


[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.2 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar dependencias

In [103]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [104]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:benjamin11@localhost:5432/tarea-bd

env: DATABASE_URL=postgresql://postgres:benjamin11@localhost:5432/tarea-bd


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [105]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='tarea-bd',
   password="benjamin11"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x00000299916B1250; dsn: 'user=postgres password=xxx dbname=tarea-bd host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [106]:
%%sql
CREATE TABLE facciones(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    descripcion VARCHAR(45) NOT NULL,
    
    PRIMARY KEY (id)

);
    

 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «facciones» ya existe

[SQL: CREATE TABLE facciones(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    descripcion VARCHAR(45) NOT NULL,
    
    PRIMARY KEY (id)

);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "trabajos" (6 pts)

In [107]:
%%sql
CREATE TABLE trabajos(
    
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    
    nombre VARCHAR(45) NOT NULL,
    
    descripcion VARCHAR(45) NOT NULL,

    sueldo INTEGER NOT NULL,
    
    PRIMARY KEY (id)

);


 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «trabajos» ya existe

[SQL: CREATE TABLE trabajos(
    
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    
    nombre VARCHAR(45) NOT NULL,
    
    descripcion VARCHAR(45) NOT NULL,

    sueldo INTEGER NOT NULL,
    
    PRIMARY KEY (id)

);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "personas" (6 pts)

In [108]:
%%sql
CREATE TABLE personas(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    apellidos VARCHAR(45) NOT NULL,

    fecha_nacimiento DATE NOT NULL,

    infectado BOOLEAN NOT NULL,

    conyugue INTEGER,

    PRIMARY KEY (id) ,
    
    FOREIGN KEY (conyugue) REFERENCES personas
);


 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «personas» ya existe

[SQL: CREATE TABLE personas(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    apellidos VARCHAR(45) NOT NULL,

    fecha_nacimiento DATE NOT NULL,

    infectado BOOLEAN NOT NULL,

    conyugue INTEGER,

    PRIMARY KEY (id) ,
    
    FOREIGN KEY (conyugue) REFERENCES personas
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "cuentas_bancarias" (6 pts)

In [109]:
%%sql

CREATE TABLE cuentas_bancarias (

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    tipo_de_cuenta VARCHAR(45) NOT NULL,

    banco VARCHAR(45) NOT NULL,

    saldo INTEGER NOT NULL,

    PRIMARY KEY (id),

    id_persona INTEGER, 

    FOREIGN KEY(id_persona) REFERENCES personas(id)

);

 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «cuentas_bancarias» ya existe

[SQL: CREATE TABLE cuentas_bancarias (

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    tipo_de_cuenta VARCHAR(45) NOT NULL,

    banco VARCHAR(45) NOT NULL,

    saldo INTEGER NOT NULL,

    PRIMARY KEY (id),

    id_persona INTEGER, 

    FOREIGN KEY(id_persona) REFERENCES personas(id)

);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "persona_pertenece_faccion" (6 pts)

In [110]:
%%sql
CREATE TABLE persona_pertenece_faccion(

    id_persona INTEGER,

    id_faccion INTEGER,

    PRIMARY KEY (id_persona, id_faccion),

    FOREIGN KEY (id_persona) REFERENCES personas (id),

    FOREIGN KEY (id_faccion) REFERENCES facciones (id)
    

);

 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «persona_pertenece_faccion» ya existe

[SQL: CREATE TABLE persona_pertenece_faccion(

    id_persona INTEGER,

    id_faccion INTEGER,

    PRIMARY KEY (id_persona, id_faccion),

    FOREIGN KEY (id_persona) REFERENCES personas (id),

    FOREIGN KEY (id_faccion) REFERENCES facciones (id)
    

);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "persona_tiene_trabajo" (6 pts)

In [111]:
%%sql
CREATE TABLE persona_tiene_trabajo (
 
 id_persona INTEGER,

 id_trabajo INTEGER,
 
 estado BOOLEAN NOT NULL,

 ultima_vez_realizado TIMESTAMP NOT NULL,   
 
 FOREIGN KEY (id_persona) REFERENCES personas (id),
 
 FOREIGN KEY (id_trabajo) REFERENCES trabajos (id),
 
 PRIMARY KEY (id_persona , id_trabajo)

);


 * postgresql://postgres:***@localhost:5432/tarea-bd
(psycopg2.errors.DuplicateTable) la relación «persona_tiene_trabajo» ya existe

[SQL: CREATE TABLE persona_tiene_trabajo (
 
 id_persona INTEGER,

 id_trabajo INTEGER,
 
 estado BOOLEAN NOT NULL,

 ultima_vez_realizado TIMESTAMP NOT NULL,   
 
 FOREIGN KEY (id_persona) REFERENCES personas (id),
 
 FOREIGN KEY (id_trabajo) REFERENCES trabajos (id),
 
 PRIMARY KEY (id_persona , id_trabajo)

);]
(Background on this error at: https://sqlalche.me/e/20/f405)


## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [112]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._


In [113]:
%%sql
/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    id ,nombre , split_part(apellidos, ' ', 1) AS primer_apellido
FROM 
    personas

 * postgresql://postgres:***@localhost:5432/tarea-bd
1000 rows affected.


[(1, 'Ines', 'Linker'),
 (2, 'Beulah', 'Hannah'),
 (3, 'Lashawn', 'Nester'),
 (5, 'Heather', 'Cook'),
 (6, 'Jean', 'West'),
 (8, 'Betty', 'Kenney'),
 (10, 'Tara', 'Smith'),
 (11, 'Maria', 'Alexander'),
 (12, 'Ruth', 'Tutas'),
 (13, 'Shelley', 'Mitchell'),
 (14, 'Jonathan', 'Willbanks'),
 (16, 'Mary', 'Kelly'),
 (17, 'Michael', 'Tidwell'),
 (19, 'Larry', 'Yeager'),
 (21, 'Christopher', 'Mcclain'),
 (24, 'Cynthia', 'Sanchez'),
 (25, 'John', 'Jones'),
 (26, 'Martha', 'Irish'),
 (28, 'Travis', 'Lawson'),
 (29, 'Hollis', 'Mode'),
 (30, 'Alica', 'Duncan'),
 (31, 'Brian', 'Robertson'),
 (32, 'Ella', 'Merritt'),
 (33, 'Katie', 'Conley'),
 (34, 'Francis', 'Binns'),
 (35, 'Millie', 'Martin'),
 (36, 'Wilda', 'Regan'),
 (37, 'Jessica', 'Jones'),
 (38, 'Robert', 'Webster'),
 (39, 'Frank', 'Sweat'),
 (40, 'Carolyn', 'Blank'),
 (41, 'Willie', 'Dunbar'),
 (42, 'Maria', 'Ayers'),
 (43, 'Leo', 'Naquin'),
 (44, 'Shirley', 'Terepka'),
 (45, 'William', 'Jones'),
 (46, 'Holly', 'Cohran'),
 (48, 'Suzanne', 'Haynes'),
 (49, 'Michael', 'Danielsen'),
 (50, 'Theresia', 'Sumstad'),
 (51, 'Stanley', 'Abron'),
 (52, 'Gloria', 'Marugg'),
 (54, 'Elinor', 'Grayson'),
 (55, 'Tracey', 'Dewees'),
 (56, 'Santos', 'Chevez'),
 (58, 'Gary', 'Viejo'),
 (59, 'Margaret', 'Larson'),
 (60, 'Franklin', 'Higdon'),
 (61, 'Raymond', 'Figueroa'),
 (62, 'Henry', 'Castilo'),
 (63, 'Betsy', 'Murphy'),
 (64, 'Lonnie', 'Fitzsimmons'),
 (65, 'John', 'Offerdahl'),
 (66, 'Annette', 'Duncan'),
 (67, 'Wendell', 'Lewis'),
 (68, 'Luke', 'Howard'),
 (69, 'Craig', 'Vaquera'),
 (70, 'Janie', 'Meza'),
 (72, 'Martine', 'Graham'),
 (74, 'Mary', 'Swayzer'),
 (75, 'Tracy', 'Mckinney'),
 (76, 'Scott', 'Piccinone'),
 (77, 'Alfredo', 'Owens'),
 (78, 'Rose', 'Barry'),
 (79, 'Kizzy', 'Derentis'),
 (80, 'Frank', 'Hogan'),
 (81, 'Katherine', 'Finch'),
 (82, 'Janel', 'Wright'),
 (83, 'Frances', 'Trujillo'),
 (84, 'Kathleen', 'Dizon'),
 (87, 'Michael', 'Washburn'),
 (88, 'Janice', 'Francis'),
 (89, 'Chad', 'Lofquist'),
 (90, 'Charlie', 'Milton'),
 (91, 'Kathy', 'Ramsey'),
 (92, 'Daniel', 'Whittier'),
 (93, 'Cindy', 'Scott'),
 (94, 'Raymond', 'Orduno'),
 (95, 'Cleo', 'Derrick'),
 (96, 'Alison', 'Jones'),
 (97, 'Victoria', 'Gilmore'),
 (98, 'Ashlee', 'Minahan'),
 (99, 'Philip', 'Winters'),
 (100, 'Anna', 'Wedderburn'),
 (101, 'Richard', 'Green'),
 (102, 'Donald', 'Murphy'),
 (105, 'Rogelio', 'Espinosa'),
 (106, 'Eloise', 'Clark'),
 (107, 'Betty', 'Self'),
 (108, 'Adam', 'Mcculler'),
 (109, 'David', 'Hester'),
 (110, 'Beatrice', 'Kirkland'),
 (111, 'Maggie', 'Stout'),
 (112, 'Roger', 'Hoffman'),
 (113, 'Gladys', 'Correla'),
 (114, 'Ronald', 'Shaw'),
 (115, 'Barbara', 'Croyle'),
 (116, 'Erica', 'Ponder'),
 (117, 'Willie', 'Matson'),
 (118, 'Nancy', 'Washington'),
 (119, 'Charles', 'Roberts'),
 (120, 'Richard', 'Cogliano'),
 (122, 'Debbie', 'Gonzalez'),
 (123, 'Donna', 'Mcbroom'),
 (124, 'Dara', 'Cogswell'),
 (125, 'Gary', 'Estevez'),
 (71, 'Josie', 'Keo'),
 (85, 'Melissa', 'Fuentes'),
 (4, 'Iris', 'Hollins'),
 (47, 'Eli', 'Williams'),
 (22, 'Adam', 'Logan'),
 (7, 'Juan', 'Sesco'),
 (86, 'Ricky', 'Wilson'),
 (15, 'Edward', 'Neal'),
 (18, 'Angela', 'Rosas'),
 (53, 'Barbara', 'Brown'),
 (57, 'Erwin', 'Gibson'),
 (9, 'Ashley', 'Duncan'),
 (121, 'Esther', 'Copeland'),
 (103, 'Ernesto', 'Smith'),
 (73, 'Brian', 'Estes'),
 (20, 'Constance', 'Gipson'),
 (126, 'Joann', 'Murphy'),
 (127, 'Patricia', 'Voegele'),
 (128, 'Tonya', 'Folse'),
 (129, 'Eric', 'Ortiz'),
 (130, 'Daniel', 'Fleming'),
 (131, 'Peggy', 'Brown'),
 (132, 'Michael', 'Rojas'),
 (135, 'Yolanda', 'Bennett'),
 (136, 'Lori', 'Jenks'),
 (137, 'Thomas', 'Janecka'),
 (138, 'Daniel', 'Hickman'),
 (139, 'Brian', 'Carter'),
 (141, 'Rosetta', 'Trotter'),
 (142, 'Kristen', 'Saxman'),
 (143, 'Gloria', 'Maybee'),
 (144, 'Seth', 'Heitz'),
 (145, 'Raymond', 'Green'),
 (146, 'Jane', 'Warner'),
 (147, 'Tanya', 'Townsend'),
 (148, 'Marshall', 'Conrad'),
 (149, 'Carmen', 'Rosenbaum'),
 (150, 'Johnny', 'Long'),
 (151, 'Thelma', 'Davis'),
 (152, 'Christina', 'Williams'),
 (

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** _Escriba aquí la descripción de su solución._

In [114]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT 

COUNT(*) AS cantidad 

FROM 

personas WHERE fecha_nacimiento > '2019-12-01';

 * postgresql://postgres:***@localhost:5432/tarea-bd
1 rows affected.


[(60,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [115]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    id_persona , nombre , split_part(apellidos, ' ', 1) AS primer_apellido , saldo AS Dinero
FROM 
    personas INNER JOIN cuentas_bancarias
ON  
    cuentas_bancarias.id_persona = personas.id




 * postgresql://postgres:***@localhost:5432/tarea-bd
120 rows affected.


[(380, 'Christopher', 'Rosboril', 6147),
 (808, 'John', 'Sanchez', 5501),
 (321, 'Adam', 'Galdames', 2478),
 (980, 'Angela', 'Brown', 6638),
 (178, 'Kenneth', 'Battles', 6671),
 (895, 'Sarah', 'Holley', 1328),
 (320, 'Deborah', 'Ricker', 5118),
 (354, 'Victor', 'Tillson', 8563),
 (143, 'Gloria', 'Maybee', 8030),
 (234, 'Karin', 'Sproule', 9713),
 (931, 'Virginia', 'Marshall', 9761),
 (942, 'Mike', 'Latney', 3816),
 (694, 'Venessa', 'Martinez', 7568),
 (286, 'Linda', 'Rodgers', 1380),
 (368, 'George', 'Pope', 4611),
 (865, 'Drew', 'Klase', 1790),
 (772, 'Matthew', 'Jones', 8094),
 (274, 'Constance', 'Robinson', 6419),
 (162, 'Ronald', 'Newman', 7897),
 (885, 'Armando', 'Lopez', 5056),
 (17, 'Michael', 'Tidwell', 8887),
 (966, 'Eileen', 'Martinez', 7054),
 (627, 'Guadalupe', 'Modafferi', 2665),
 (437, 'Julian', 'Harris', 7942),
 (350, 'Cornelia', 'Guest', 9567),
 (196, 'Tasha', 'Stoner', 9570),
 (775, 'Catherine', 'Jones', 6842),
 (972, 'Khadijah', 'Rodriguez', 753),
 (386, 'Christina', 'Hunt', 3227),
 (131, 'Peggy', 'Brown', 8174),
 (77, 'Alfredo', 'Owens', 3619),
 (102, 'Donald', 'Murphy', 9703),
 (813, 'Casey', 'Johnson', 2565),
 (112, 'Roger', 'Hoffman', 5496),
 (94, 'Raymond', 'Orduno', 5064),
 (738, 'Casey', 'Griffin', 9742),
 (784, 'Tina', 'Li', 7850),
 (468, 'Calvin', 'Mulholland', 3783),
 (21, 'Christopher', 'Mcclain', 6689),
 (844, 'William', 'Terry', 1756),
 (463, 'Ashley', 'Knight', 549),
 (639, 'Irena', 'Geter', 621),
 (200, 'Edwin', 'Penn', 8217),
 (545, 'Richard', 'Mills', 1961),
 (408, 'Jean', 'Christensen', 1508),
 (851, 'Michaela', 'Thomas', 3170),
 (358, 'Rachael', 'Walker', 5816),
 (564, 'Jean', 'Guzman', 804),
 (938, 'Mark', 'Hickman', 6749),
 (525, 'William', 'Brule', 7707),
 (510, 'Lee', 'Selfe', 5202),
 (46, 'Holly', 'Cohran', 4490),
 (288, 'Michelle', 'Nee', 635),
 (866, 'Christopher', 'Andrews', 9075),
 (10, 'Tara', 'Smith', 81),
 (971, 'Lynne', 'Scipio', 7781),
 (831, 'Tyler', 'Hansell', 131),
 (588, 'Charlotte', 'Goldstein', 7844),
 (403, 'Oliva', 'Day', 3551),
 (757, 'Mitchell', 'Quinn', 6684),
 (955, 'Jack', 'Satterwhite', 9483),
 (338, 'Bruce', 'Smart', 1585),
 (745, 'Nada', 'Belcher', 4189),
 (669, 'Arthur', 'Nehlsen', 7000),
 (947, 'Raymond', 'Ringo', 7691),
 (649, 'Preston', 'Hernandez', 9352),
 (999, 'Ethel', 'Marshall', 3246),
 (693, 'Gwendolyn', 'Capley', 9566),
 (901, 'Scott', 'Lavigne', 6393),
 (341, 'Linda', 'Abramowski', 433),
 (783, 'Nicholas', 'Tallent', 9967),
 (72, 'Martine', 'Graham', 9678),
 (746, 'Robert', 'Hayes', 1963),
 (422, 'John', 'Herd', 1515),
 (667, 'John', 'Mathis', 9389),
 (504, 'Ronald', 'White', 7931),
 (338, 'Bruce', 'Smart', 8731),
 (823, 'Cecilia', 'Mason', 1237),
 (94, 'Raymond', 'Orduno', 6313),
 (532, 'James', 'Peterson', 7485),
 (825, 'Sharon', 'Demarest', 4934),
 (433, 'Milford', 'Lane', 3043),
 (677, 'Nicole', 'Leatherwood', 1358),
 (280, 'Ebony', 'Wiggins', 4670),
 (181, 'Valerie', 'Dreher', 9366),
 (624, 'Linda', 'Ezell', 2053),
 (738, 'Casey', 'Griffin', 9015),
 (890, 'Israel', 'Earle', 3005),
 (602, 'Michael', 'Gourley', 5638),
 (631, 'Robert', 'Deleon', 5813),
 (709, 'Howard', 'Frazer', 5821),
 (970, 'Ricardo', 'Crawford', 45),
 (400, 'Lupe', 'Little', 1090),
 (164, 'Shirley', 'Palovick', 6809),
 (252, 'Mary', 'Graff', 9278),
 (304, 'Mammie', 'Molina', 7854),
 (992, 'Carrie', 'Willier', 1303),
 (672, 'Laticia', 'Wang', 9950),
 (702, 'Kimberly', 'Pichler', 9536),
 (506, 'Eileen', 'Perez', 9793),
 (810, 'Richard', 'Smith', 4784),
 (981, 'William', 'Moore', 5770),
 (720, 'Linda', 'Guzman', 7877),
 (663, 'William', 'Means', 1942),
 (939, 'Christy', 'Lovins', 6445),
 (794, 'Kelly', 'Angles', 5028),
 (477, 'Herman', 'Montgomery', 557),
 (281, 'Caitlin', 'Ramey', 9279),
 (125, 'Gary', 'Estevez', 953),
 (974, 'Stacey', 'Pedroza', 7215),
 (519, 'Celeste', 'Gober', 9151),
 (147, 'Tanya', 'Townsend', 5670),
 (797, 'Thomas', 'Maier', 6007),
 (750, 'Mark', 'Velazquez', 9179),
 (463, 'Ashley', 'Knight', 7130),
 (650, 'Kelly', 'Thomas', 3270),
 (451, 'John', 'Pleasanton

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [132]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT facciones.nombre AS faccion, COUNT(persona_tiene_trabajo.id_persona) AS cantidad_trabajadores_activos
FROM facciones 
LEFT JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
LEFT JOIN persona_tiene_trabajo persona_tiene_trabajo ON persona_pertenece_faccion.id_persona = persona_tiene_trabajo.id_persona
WHERE persona_tiene_trabajo.estado = true
GROUP BY facciones.nombre;

    

 * postgresql://postgres:***@localhost:5432/tarea-bd
6 rows affected.


[('JavaScriptia', 10),
 ('Luciérnagas', 9),
 ('Pythonia', 9),
 ('MaCdonia', 11),
 ('FEDRA', 5),
 ('Carroñeros', 13)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    * 
FROM 
    *

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    * 
FROM 
    *

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *